conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/ConjureExceptionHandlerTest.java

```diff
@@ -141,64 +137,65 @@ public void handles403RemoteException() throws IOException {
         ByteArrayOutputStream stream = new ByteArrayOutputStream();
         Encodings.json().serializer(new TypeMarker<SerializableError>() {}).serialize(expectedPropagatedError, stream);
 
-        assertThat(response.body().string()).isEqualTo(stream.toString(StandardCharsets.UTF_8));
-        assertThat(response.code()).isEqualTo(ErrorType.PERMISSION_DENIED.httpErrorCode());
+        assertThat(connection.getResponseCode()).isEqualTo(ErrorType.PERMISSION_DENIED.httpErrorCode());
+        assertThat(getErrorBody(connection)).isEqualTo(stream.toString(StandardCharsets.UTF_8));
     }
 
     @Test
     public void handlesQosExceptionThrottleWithoutDuration() throws IOException {
         exception = QosException.throttle();
-        Response response = execute();
+        HttpURLConnection connection = execute();
 
-        assertThat(response.code()).isEqualTo(429);
-        assertThat(response.body().string()).isEmpty();
-        assertThat(response.headers().toMultimap()).doesNotContainKey("Retry-After");
-        assertThat(response.headers().toMultimap()).containsOnlyKeys("connection", "content-length", "date");
+        assertThat(connection.getResponseCode()).isEqualTo(429);
+        assertThat(connection.getErrorStream()).isNull();
+        assertThat(connection.getHeaderFields()).doesNotContainKey("Retry-After");
+        assertThat(connection.getHeaderFields()).containsKeys("Connection", "Content-Length", "Date");
```

Comment:
Note that the header maps behave a little differently with `HttpUrlConnection` in two ways, they are capitalized and they have a null entry which makes the `containsOnlyKeys` fail. Technically this makes the test less resilient to other random headers. I could put a null in to the var args and preserve the method if you want.

Reply:
I don't think we should fail the tests if other headers are set, that's a bit fragile from a protocol perspective anyhow :+1: